data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/leader/LeaderScheduler.java

```diff
@@ -0,0 +1,344 @@
+
+    private static final Logger LOG = LoggerFactory.getLogger(LeaderScheduler.class);
+
+    /**
+     * Default duration to extend the timeout of lease
+     */
+    private static final int DEFAULT_EXTEND_LEASE_MINUTES = 3;
+
+    /**
+     * Default interval to run lease check and shard discovery
+     */
+    private static final int DEFAULT_LEASE_INTERVAL_MILLIS = 60_000;
+
+    private final List<TableConfig> tableConfigs;
+
+    private final EnhancedSourceCoordinator coordinator;
+
+    private final DynamoDbClient dynamoDbClient;
+
+    private final ShardManager shardManager;
+
+    private LeaderPartition leaderPartition;
+
+    private List<String> streamArns;
+
+    public LeaderScheduler(EnhancedSourceCoordinator coordinator, DynamoDbClient dynamoDbClient, ShardManager shardManager, List<TableConfig> tableConfigs) {
+        this.tableConfigs = tableConfigs;
+        this.coordinator = coordinator;
+        this.dynamoDbClient = dynamoDbClient;
+        this.shardManager = shardManager;
+    }
+
+    @Override
+    public void run() {
+        LOG.debug("Starting Leader Scheduler for initialization and shard discovery");
+
+        while (!Thread.currentThread().isInterrupted()) {
+            try {
+                // Try to acquire the lease if not owned.
+                if (leaderPartition == null) {
+                    final Optional<EnhancedSourcePartition> sourcePartition = coordinator.acquireAvailablePartition(LeaderPartition.PARTITION_TYPE);
+                    if (sourcePartition.isPresent()) {
+                        LOG.info("Running as a LEADER node");
+                        leaderPartition = (LeaderPartition) sourcePartition.get();
+
+                        LeaderProgressState leaderProgressState = leaderPartition.getProgressState().get();
+                        if (!leaderProgressState.isInitialized()) {
+                            LOG.debug("The service is not been initialized");
+                            init();
+                        }
+
+                        // The initialization process will populate that value, otherwise, get from state
+                        if (streamArns == null) {
+                            streamArns = leaderProgressState.getStreamArns();
+                        }
+                    }
+                }
+                // Once owned, run Normal LEADER node process.
+                if (leaderPartition != null) {
+                    // Run shard discovery,
+                    // and then find children for all completed shards in the last 1 day (maximum time)
+                    // A shard will be expired after 24 hours.
+                    discoveryAndCreateChildrenPartitions();
+                    // Extend the timeout
+                    coordinator.saveProgressStateForPartition(leaderPartition, Duration.ofMinutes(DEFAULT_EXTEND_LEASE_MINUTES));
+
+                }
+
+            } catch (Exception e) {
+                e.printStackTrace();
```

Comment:
Please handle this differently. Either re-throw or log it using SLF4J.

Reply:
Sure, will make the change.